File: src/features/scheduler/domain/scheduler-input.ts
import { CompanyMember } from "@/features/auth/domain/types";
import { buildDayOverviews } from "./calendar";
import { MemberAvailability, SchedulerInput, SchedulerMember, SchedulerShiftRequest } from "./types";

function sortUnique(values: string[]) {
  return Array.from(new Set(values)).sort((left, right) => left.localeCompare(right));
}

export function buildSchedulerMembers(
  companyMembers: CompanyMember[],
  availabilities: MemberAvailability[],
): SchedulerMember[] {
  const availabilityByMember = new Map(
    availabilities.map((availability) => [
      availability.memberId,
      sortUnique(availability.entries.map((entry) => entry.date)),
    ]),
  );

  const membersWithAvailability = new Set(availabilities.map((availability) => availability.memberId));

  return companyMembers
    .filter((member) => membersWithAvailability.has(member.userId))
    .map((member) => ({
      memberId: member.userId,
      name: member.name,
      email: member.email,
      active: member.schedulingProfile.active,
      maxShifts: member.schedulingProfile.maxShifts,
      preferredCoworkerIds: sortUnique(member.schedulingProfile.preferredCoworkerIds),
      availabilityDates: availabilityByMember.get(member.userId) ?? [],
    }))
    .sort((left, right) => left.name.localeCompare(right.name));
}

export function buildSchedulerShiftRequests(
  monthId: string,
  intakeLimitPerShift: number,
  availabilities: MemberAvailability[],
): SchedulerShiftRequest[] {
  return buildDayOverviews(monthId, intakeLimitPerShift, availabilities).flatMap((day) =>
    day.shifts.map((shift) => ({
      shiftId: `${day.date}:${shift.shiftType}`,
      date: day.date,
      shiftType: shift.shiftType,
      capacity: shift.capacity,
      candidateMemberIds: shift.candidateIds.slice(),
    })),
  );
}

export function buildSchedulerInput(input: {
  monthId: string;
  intakeLimitPerShift: number;
  availabilities: MemberAvailability[];
  companyMembers: CompanyMember[];
}): SchedulerInput {
  return {
    monthId: input.monthId,
    members: buildSchedulerMembers(input.companyMembers, input.availabilities),
    shiftRequests: buildSchedulerShiftRequests(input.monthId, input.intakeLimitPerShift, input.availabilities),
  };
}

File: backend/app/domain/scheduler_input.py
from __future__ import annotations

from app.domain.calendar import build_day_overviews
from app.domain.types import (
    TeamMember,
    MemberAvailability,
    SchedulerInput,
    SchedulerMember,
    SchedulerShiftRequest,
)


def build_scheduler_members(
    team_members: list[TeamMember],
    availabilities: list[MemberAvailability],
) -> list[SchedulerMember]:
    availability_dates_by_member = {
        a.member_id: sorted(set(e.date for e in a.entries))
        for a in availabilities
    }
    members_with_availability = set(a.member_id for a in availabilities)

    result = [
        SchedulerMember(
            member_id=m.user_id,
            name=m.name,
            email=m.email,
            active=m.active,
            max_shifts=m.max_shifts,
            preferred_coworker_ids=sorted(set(m.preferred_coworker_ids)),
            availability_dates=availability_dates_by_member.get(m.user_id, []),
        )
        for m in team_members
        if m.user_id in members_with_availability
    ]
    return sorted(result, key=lambda m: m.name)


def build_scheduler_shift_requests(
    month_id: str,
    intake_limit_per_shift: int,
    availabilities: list[MemberAvailability],
) -> list[SchedulerShiftRequest]:
    return [
        SchedulerShiftRequest(
            shift_id=f"{day.date}:{shift.shift_type}",
            date=day.date,
            shift_type=shift.shift_type,
            capacity=shift.capacity,
            candidate_member_ids=list(shift.candidate_ids),
        )
        for day in build_day_overviews(month_id, intake_limit_per_shift, availabilities)
        for shift in day.shifts
    ]


def build_scheduler_input(
    month_id: str,
    intake_limit_per_shift: int,
    availabilities: list[MemberAvailability],
    team_members: list[TeamMember],
) -> SchedulerInput:
    return SchedulerInput(
        month_id=month_id,
        members=build_scheduler_members(team_members, availabilities),
        shift_requests=build_scheduler_shift_requests(month_id, intake_limit_per_shift, availabilities),
    )
